src/Test/Plutip/Contract.hs

```diff
@@ -113,176 +160,70 @@ type TestContractConstraints (w :: Type) (e :: Type) (a :: Type) =
-
- @since 0.2
--}
-assertObservableStateWith ::
-  forall (w :: Type) (e :: Type) (a :: Type).
-  TestContractConstraints w e a =>
-  String ->
-  TestWallets ->
-  (w -> Bool) ->
-  TestRunner w e a ->
-  (TestWallets, IO (ClusterEnv, NonEmpty BpiWallet) -> TestTree)
-assertObservableStateWith tag testWallets predicate testRunner =
-  ( testWallets
-  , singleTest tag
-      . TestContract
-        testRunner
-        (ExpectSuccess (const True) predicate (twExpected <$> unTestWallets testWallets))
-  )
-
-{- | Check if the observable state of the contract equals to some expected value.
-
- @since 0.2
--}
-shouldHaveObservableState ::
-  forall (w :: Type) (e :: Type) (a :: Type).
-  (TestContractConstraints w e a, Eq w) =>
-  String ->
-  TestWallets ->
-  w ->
-  TestRunner w e a ->
-  (TestWallets, IO (ClusterEnv, NonEmpty BpiWallet) -> TestTree)
-shouldHaveObservableState tag testWallets expected =
-  assertObservableStateWith tag testWallets (== expected)
-
-{- | Check if contract throws expected error.
-
- @since 0.2
--}
-shouldThrowContractError ::
-  forall (w :: Type) (e :: Type) (a :: Type).
-  (TestContractConstraints w e a, Eq e) =>
-  String ->
-  TestWallets ->
-  e ->
-  TestRunner w e a ->
-  (TestWallets, IO (ClusterEnv, NonEmpty BpiWallet) -> TestTree)
-shouldThrowContractError tag testWallets expected =
-  assertContractError tag testWallets (== expected)
-
-{- | Assert the error thrown by contract with a custom predicate.
+assertExecution tag testWallets testRunner predicates =
+  (testWallets, toTestGroup)
+  where
+    toTestGroup ioEnv =
+      withResource (runReaderT testRunner =<< ioEnv) (const $ pure ()) $
+        \ioRes ->
+          testGroup tag $
+            maybeAddValuesCheck
+              ioRes
+              testWallets
+              (toCase ioRes <$> predicates)
+
+    -- wraps IO with result of contract execution into single test
+    toCase ioRes p =
+      singleTest (pTag p) (TestContract p ioRes)
+
+{- | Adds test case with assertions on values if any assertions were added
+  by `initAndAssert...` functions during wallets setup
 
  @since 0.2
 -}
-assertContractError ::
-  forall (w :: Type) (e :: Type) (a :: Type).
-  (TestContractConstraints w e a) =>
-  String ->
+maybeAddValuesCheck ::
+  Show e =>
+  IO (ExecutionResult w e (a, NonEmpty Value)) ->
   TestWallets ->
-  (e -> Bool) ->
-  TestRunner w e a ->
-  (TestWallets, IO (ClusterEnv, NonEmpty BpiWallet) -> TestTree)
-assertContractError tag testWallets predicate =
-  assertFailure tag testWallets (mkPredicate predicate)
+  [TestTree] ->
+  [TestTree]
+maybeAddValuesCheck ioRes tws =
+  bool id (valuesCheckCase :) (any isJust expected)
   where
-    mkPredicate p = \case
-      ContractExecutionError e -> p e
-      _ -> False
+    expected = twExpected <$> unTestWallets tws
 
-{- | Assert the failure reason of contract execution with a custom predicate.
+    valuesCheckCase = testCase "Values check" $ do
```

Comment:
I think we don't need `do` notation here, but for readability it'd be even better to replace `>>=` by something like `ExecutionResult o _ <- ioRes` (but this is more like a personal preference)